### PR TITLE
Changes to support config-setup service for multi-npu

### PR DESCRIPTION
**Pull request by abdosi**
**- Why I did it**
Changes to support config-setup service for multi-npu platforms.
we are using python commands config reload/save/load-minigraph 
so that there is common code. when being called from this script
we will not stop/start docker services. This PR is related to this:
https://github.com/Azure/sonic-utilities/pull/919
For Multi-npu we are not supporting as of
now config initialization and ZTP. It will support creating
config db from minigraph if present or using  config db from previous
file system

Submodule update PR for sonic-utilities is : https://github.com/Azure/sonic-buildimage/pull/4622
Signed-off-by: Abhishek Dosi <abdosi@microsoft.com>

**- How I did it**
Updated the service to handle multiple config db in case of multi-npu platforms

**- How to verify it**
Manual testing by running /usr/bin/config-setup on both single and multi-npu
platforms.

### CODE REVIEW

File: files/image_config/config-setup/config-setup [hunk 1]
```diff
@@ -281,15 +380,51 @@ copy_post_migration_hooks()
...
+    do
+        config_db_file_list+=' '${CONFIG_DB_PREFIX}$asic_num${CONFIG_DB_SUFFIX}
+        ((asic_num = asic_num + 1))
+    done
+
+    echo $config_db_file_list
+}
+# Check if all needed config db are prsesnt for both
+# single and multi-npu platforms
+check_config_db_present()
```
Comment (yxieca) [inline files/image_config/config-setup/config-setup:399]: suggest rename this function to check_all_config_db_present() to match its return values.

File: files/image_config/config-setup/config-setup [hunk 2]
```diff
@@ -281,15 +380,51 @@ copy_post_migration_hooks()
...
+{
+    if [[ ! -r ${CONFIG_DB_JSON} ]]; then
+       return 1
+    fi
+    asic_num=0
+    while [[ ($asic_num -lt $NUM_ASIC) && ($NUM_ASIC -gt 1) ]]
+    do
```
Comment (yxieca) [inline files/image_config/config-setup/config-setup:406]: Very minor issue. Suggest using "while [[...]]; do" style to match other code style in this file. e.g. "if [[...]]; then"

File: files/image_config/config-setup/config-setup [hunk 3]
```diff
@@ -373,6 +509,16 @@ boot_config()
 }
 
 ### Execution starts here ###
+PLATFORM=`sonic-cfggen -H -v DEVICE_METADATA.localhost.platform`
+# Parse the device specific asic conf file, if it exists
+ASIC_CONF=/usr/share/sonic/device/$PLATFORM/asic.conf
+if [ -f "$ASIC_CONF" ]; then
+source $ASIC_CONF
```
Comment (yxieca) [inline files/image_config/config-setup/config-setup:516]: Minor issue: indentation.

File: files/image_config/config-setup/config-setup [hunk 4]
```diff
@@ -99,28 +103,123 @@ run_hookdir() {
...
+    elif [ "$1" == "reload" ]; then
+        # case when config db is already present from previous image file system
+        sonic-cfggen ${NAMESPACE_ARGUMENT} -j ${CONFIG_DB_PATH}${CONFIG_DB_PREFIX}$2${CONFIG_DB_SUFFIX} --write-to-db
+    fi
+
+    return 0
+}
+configdb_migrator()
+{
+    if [[ ! -z "$2" ]]; then
```
Comment (yxieca) [inline files/image_config/config-setup/config-setup:138]: you could use -n instead of '! -z'

File: files/image_config/config-setup/config-setup [hunk 5]
```diff
@@ -99,28 +103,123 @@ run_hookdir() {
...
+        NAMESPACE_ARGUMENT=${NAMESPACE_OPTION}${NAMESPACE_PREFIX}$2
+    else
+        NAMESPACE_ARGUMENT=$2
```
Comment (yxieca) [inline files/image_config/config-setup/config-setup:141]: suggest explicitly assign NAMESPACE_ARGUMENT to empty string.

File: files/image_config/config-setup/config-setup (hunk 4 above)
Comment (yxieca) [reply files/image_config/config-setup/config-setup:138]: Or maybe checking $# here is more efficient? You do need $# >= 1, right?

File: files/image_config/config-setup/config-setup (hunk 5 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @yxieca Done

File: files/image_config/config-setup/config-setup (hunk 4 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @yxieca Thanks for pointing this. Done

File: files/image_config/config-setup/config-setup (hunk 3 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @yxieca Done

File: files/image_config/config-setup/config-setup (hunk 2 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @yxieca Done

File: files/image_config/config-setup/config-setup (hunk 1 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @yxieca Done

File: files/image_config/config-setup/config-setup [hunk 6]
```diff
@@ -99,28 +103,119 @@ run_hookdir() {
     return $exit_status
 }
 
+initalize_configdb()
```
Comment (lguohan) [inline files/image_config/config-setup/config-setup:106]: can instead call config reload -y instead of implementing here? it is code duplication.

Comment (abdosi) [reply files/image_config/config-setup/config-setup:106]: @lguohan I was thinking other way when making change. config reload/save command does stop/start services which is not needed during config-setup script. 
We can update those commands to call config-setup shell script with maintaining stop/start of services.
That way code will be present only in config-setup for config db generation.

File: files/image_config/config-setup/config-setup [hunk 7]
```diff
@@ -281,15 +376,49 @@ copy_post_migration_hooks()
...
+    while [[ ($asic_num -lt $NUM_ASIC) && ($NUM_ASIC -gt 1) ]]; do
+        config_db_file_list+=' '${CONFIG_DB_PREFIX}$asic_num${CONFIG_DB_SUFFIX}
+        ((asic_num = asic_num + 1))
+    done
+
+    echo $config_db_file_list
+}
+# Check if all needed config db are prsesnt for both
+# single and multi-npu platforms
+check_all_config_db_present()
```
Comment (judyjoseph) [inline files/image_config/config-setup/config-setup:394]: May be we could use the API get_config_db_file_list() to get the file list as a string and loop through to check if they exists ?

File: files/image_config/config-setup/config-setup [hunk 8]
```diff
@@ -99,28 +103,119 @@ run_hookdir() {
...
+        # case when config db is already present from previous image file system
+        sonic-cfggen ${NAMESPACE_ARGUMENT} -j ${CONFIG_DB_PATH}${CONFIG_DB_PREFIX}$2${CONFIG_DB_SUFFIX} --write-to-db
+    fi
+
+    return 0
+}
+configdb_migrator()
+{
+    NAMESPACE_ARGUMENT=''
+    IP_NETNS_CMD=''
```
Comment (judyjoseph) [inline files/image_config/config-setup/config-setup:138]: this variable is not used here

File: files/image_config/config-setup/config-setup [hunk 9]
```diff
@@ -99,28 +103,119 @@ run_hookdir() {
...
+        # need to be migrated
+        if [[ -x /usr/bin/db_migrator.py ]]; then
+            # Migrate the DB to the latest schema version if needed
+            /usr/bin/db_migrator.py ${NAMESPACE_ARGUMENT} -o migrate
+        fi
+    fi
+}
+
 # Reload minigraph.xml file on disk
 reload_minigraph()
```
Comment (judyjoseph) [inline files/image_config/config-setup/config-setup:161]: There is a reload_minigraph() in the updategraph service in  files/image_config/updategraph/updategraph. 
That needs to be modified in a similar way ?

File: files/image_config/config-setup/config-setup (hunk 7 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup:394]: get_config_db_file_list is just file name and the complete path .  It is used to copy file from old_config. check_all_config_db_present is doing loop instead

File: files/image_config/config-setup/config-setup (hunk 8 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup:138]: @judyjoseph thanks will remove it

File: files/image_config/config-setup/config-setup (hunk 9 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup:161]: @judyjoseph for multi-npu platforms we are not supporting getting minigraph from DHCP service and doing reload as of now. That will be in TO-DO list

**Comment by yxieca** [comment]
You have a mixture of [...] and [[...]]. Suggest unify to [[....]]

**Comment by abdosi** [comment]
> You have a mixture of [...] and [[...]]. Suggest unify to [[....]]

@yxieca Done for new changes. However there are many places where we have this mismatch so we might need to revisit for complete update.

File: files/image_config/config-setup/config-setup (hunk 8 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @judyjoseph Done

File: files/image_config/config-setup/config-setup (hunk 6 above)
Comment (lguohan) [reply files/image_config/config-setup/config-setup:106]: the reason the other way is preferred is that we can add more unit tests for the python script, but not bash script.

Comment (abdosi) [reply files/image_config/config-setup/config-setup:106]: @lguohan. Python script will still exist but internally will call shell script command. Here is the reference PR: https://github.com/Azure/sonic-utilities/pull/919

Comment (abdosi) [reply files/image_config/config-setup/config-setup]: @Iguohan Updated to use python commands.

File: files/image_config/config-setup/config-setup (hunk 9 above)
Comment (abdosi) [reply files/image_config/config-setup/config-setup:107]: @judyjoseph updated updategraph also

**Comment by abdosi** [comment]
retest vsimage please

**Comment by abdosi** [comment]
retest mellanox please

**Comment by abdosi** [comment]
@lguohan Can you please review the PR.

**Pull request merged**